File: Linked list/linked-list.py
from typing import Any

class Node:

    value = None
    prev = None
    next = None

    def __init__(self, value, prev = None, next = None) -> None:
        self.value = value
        self.prev = prev
        self.next = next

class LinkedList:

    head: Node = None
    tail: Node = None
    length: int = 0

    def __init__(self, value: int = None) -> None:
        if value != None:
            self.head = Node(value, None, None)
            self.length = 1
        else:
            self.head = None
            self.length = 0
        self.tail = None

    def add(self, value: int) -> None:
        if self.head == None:
            self.head = Node(value, None, None)
        elif self.tail == None:
            self.tail = Node(value, self.head, None)
            self.head.next = self.tail
        else:
            self.tail.next = Node(value, self.tail, None)
            self.tail = self.tail.next
        self.length += 1

    def shift(self) -> None:
        self.head = self.head.next
        self.length -= 1

    def pop(self) -> None:
        if self.length == 1:
            self.shift()
        else:
            self.tail.prev = None
            self.tail = self.tail.prev
            self.length -= 1

    def remove(self, n: int) -> None:
        if n < 0:
            n = self.length + n
        if self.length == 0:
            raise "List is empty"
        if n >= self.length:
            raise "Index is bigger than length"
        if n == 0:
            self.shift()
        elif n == self.length - 1:
            self.pop()
        elif n < self.length / 2:
            node = self.head
            for _ in range(0, n):
                node = node.next
            node.next.prev = node.prev
            node.prev.next = node.next
            self.length -= 1
        else:
            node = self.tail
            for _ in range(self.length - 1, n, -1):
                node = node.prev
            node.next.prev = node.prev
            node.prev.next = node.next
            self.length -= 1
    
    def get_node(self, n: int) -> Node:
        if n < 0:
            n = self.length + n
        if self.length == 0:
            raise "List is empty"
        if n >= self.length:
            raise "Index is bigger than length"
        if n == 0:
            return self.head
        elif n == self.length - 1:
            return self.tail
        elif n < self.length / 2:
            node = self.head
            for _ in range(0, n):
                node = node.next
            return node
        else:
            node = self.tail
            for _ in range(self.length - 1, n, -1):
                node = node.prev
            return node
    
    def get(self, n: int) -> Any:
        return self.get_node(n)

File: Linked list/linked-list.go
package main

type Node[T any] struct {
	Value T
	Prev  *Node[T]
	Next  *Node[T]
}

type LinkedList[T any] struct {
	Head   *Node[T]
	Tail   *Node[T]
	Length uint
}

func NewEmptyLinkedList[T any]() LinkedList[T] {
	return LinkedList[T]{nil, nil, 0}
}

func NewLinkedList[T any](value T) LinkedList[T] {
	return LinkedList[T]{&Node[T]{value, nil, nil}, nil, 1}
}

func (ll *LinkedList[T]) Add(value T) {
	if ll.Head == nil {
		ll.Head = &Node[T]{value, nil, nil}
	} else if ll.Tail == nil {
		ll.Tail = &Node[T]{value, ll.Head, nil}
		ll.Head.Next = ll.Tail
	} else {
		ll.Tail.Next = &Node[T]{value, ll.Tail, nil}
		ll.Tail = ll.Tail.Next
	}
	ll.Length++
}

func (ll *LinkedList[T]) Shift() {
	ll.Head = ll.Head.Next
	ll.Length--
}

func (ll *LinkedList[T]) Pop() {
	if ll.Length == 1 {
		ll.Shift()
	} else {
		ll.Tail.Prev.Next = nil
		ll.Tail = ll.Tail.Prev
		ll.Length--
	}
}

func (ll *LinkedList[T]) Remove(n int) {
	if n < 0 {
		n = int(ll.Length) + n
	}
	if ll.Length == 0 {
		panic("List is empty")
	}
	if n >= int(ll.Length) {
		panic("Index is bigger than length")
	}
	if n == 0 {
		ll.Shift()
	} else if n == int(ll.Length)-1 {
		ll.Pop()
	} else if n < int(ll.Length)/2 {
		node := ll.Head
		for i := 0; i < n; i++ {
			node = node.Next
		}
		node.Next.Prev = node.Prev
		node.Prev.Next = node.Next
		ll.Length--
	} else {
		node := ll.Tail
		for i := int(ll.Length) - 1; i > n; i-- {
			node = node.Prev
		}
		node.Next.Prev = node.Prev
		node.Prev.Next = node.Next
		ll.Length--
	}
}

func (ll *LinkedList[T]) GetNode(n int) *Node[T] {
	if n < 0 {
		n = int(ll.Length) + n
	}
	if ll.Length == 0 {
		panic("List is empty")
	}
	if n >= int(ll.Length) {
		panic("Index is bigger than length")
	}
	if n == 0 {
		return ll.Head
	} else if n == int(ll.Length)-1 {
		return ll.Tail
	} else if n < int(ll.Length)/2 {
		node := ll.Head
		for i := 0; i < n; i++ {
			node = node.Next
		}
		return node
	} else {
		node := ll.Tail
		for i := int(ll.Length) - 1; i > n; i-- {
			node = node.Prev
		}
		return node
	}
}

func (ll *LinkedList[T]) Get(n int) T {
	return ll.GetNode(n).Value
}

func main() {
}
